### PR TITLE
ENH: linalg.eig: support batched input

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Follow-up to gh-21446
Toward gh-21466

#### What does this implement/fix?
This adds support for n-dimensional batch input to `linalg.eig`.

#### Additional information
~~I'll write more to motivate this PR shortly.~~ (Done - see gh-21466). A few things to keep in mind:

- Addition of batch support in compiled code would provide speed and memory improvements, but this will take some time. In the meantime, the decorator introduced in this PR can help keep the `linalg` API more self-consistent and consistent with the NumPy API.
- I've included a test for `eig`, but I would really appreciate it if someone else would be willing to write a more general property-based test. In `stats`, we have standard tests for the `axis`, `nan_policy`, and `keepdims` arguments that are more thorough than ad hoc tests previously included for these features.
- The decorator can also add support for array-API compatible arrays if desired. In `stats`, we've also used a decorator to help standardize documentation and input validation. In `linalg`, it might be useful to standardize notes about common arguments like `overwrite_` and `check_finite` and implement standard shape/dtype checks.

### CODE REVIEW

File: scipy/linalg/_decomp.py
Comment (mdhaber) [inline scipy/linalg/_decomp.py:131]: This doesn't need to be added manually.

File: scipy/linalg/tests/test_decomp.py
Comment (mdhaber) [inline scipy/linalg/tests/test_decomp.py:409]: I need to play with `overwrite_a`/`overwrite_b` manually - I haven't thought carefully about them yet, but it might be possible to make them work for the N-dimensional arrays in some cases. That might just have to wait for lower-level implementation, though.

File: scipy/_lib/_util.py
Comment (mdhaber) [inline scipy/_lib/_util.py:1029]: I can `try`/`except` and re-raise with a tailored message, if desired.

File: scipy/_lib/_util.py
Comment (mdhaber) [inline scipy/_lib/_util.py:971]: I tried to keep this very simple/minimal for starters.

File: scipy/_lib/_util.py
Comment (ilayn) [inline scipy/_lib/_util.py:1038]: I did not perform any measurements but this step, the `list(zip(...))` and `np.stack` below will make this less performant and memory hungry since you will be holding all the nD-slices and the growing `result` in memory simultaneously until it finishes and you will be pushing a growing array to the garbage collector which can start affecting performance during the computation.

A preallocation of `result` and placing the results to each slice of the `result` is a less wasteful solution in my opinion. I did a similar thing in `expm` as follows (basically looping over the slices in an ndarray). 

https://github.com/scipy/scipy/blob/04ec412b9d990cf2bfa3704e769b6a4d185c799a/scipy/linalg/_matfuncs.py#L312-L313

If it is not nD but just 2D, the loop spins only once.

Comment (ilayn) [reply scipy/_lib/_util.py:1038]: Still not ideal since these are much less wasteful to do in Cython or C because LAPACK calls then performed inplace but until then we can live with it.

File: scipy/linalg/_decomp.py
Comment (rgommers) [reply scipy/linalg/_decomp.py:131]: I think I'd prefer regular updates of the parameter description over this note. As an example, here is what [torch.linalg.eig](https://pytorch.org/docs/main/generated/torch.linalg.eig.html) uses for the first input parameter:

_A ([Tensor](https://pytorch.org/docs/main/tensors.html#torch.Tensor)) – tensor of shape (*, n, n) where * is zero or more batch dimensions consisting of diagonalizable matrices._

That will read nicer (it's just a feature, not a special case treated with a `.. note::` admonition), and also gives the space to clarify behavior better. For example for this function: if `a` has batch dimensions, must `b` have them too with matching shape? What happens if `b` isn't batched, does it broadcast or raise?

Comment (ilayn) [reply scipy/linalg/_decomp.py:131]: The matrices are not necessarily diagonalizable though. Not sure why they added that detail. It can have jordan blocks, (not necessarily computed accurately or computed at all but it does not require diagonalizability to be a tensor).

File: scipy/_lib/_util.py
Comment (mdhaber) [reply scipy/_lib/_util.py:1038]: Yes, I considered pre-allocating. It would have made it a bit more complicated because I would have to call the function once to figure out the result type, pre-allocate, then do the loop.

Also, we don't want this to be *too* efficient, right? Need some motivation to come back and improve the efficiency with compiled code, right? : )

Anyway, if you want me to make it a little more complex for that performance bump, I will. I was just trying to start as simple as possible.

File: scipy/linalg/_decomp.py
Comment (mdhaber) [reply scipy/linalg/_decomp.py:131]: > For example for this function: if a has batch dimensions, must b have them too with matching shape? What happens if b isn't batched, does it broadcast or raise?

For all functions, the decorator broadcasts the  batch shapes, ignoring the core dimensions. The concepts here would be covered in the linked tutorial. 

> That will read nicer

It will read *nicer* than a description in terms of only the core dimensions?

If this is very important, I will open a separate issue to discuss because there's a bit more to be said.

Of course, I would be happy to change the location of a standard note.

Let me know if I should open that issue.

File: scipy/_lib/_util.py
Comment (ilayn) [reply scipy/_lib/_util.py:1038]: Certainly, but it shouldn't be too bad in my opinion compared to user is calling eig in a loop and doing it themselves. Especially when we have a chance to do it relatively OK fashion.

Comment (mdhaber) [reply scipy/_lib/_util.py:1046]: OK, I'll add this in the next commit.

File: scipy/linalg/_decomp.py
Comment (mdhaber) [reply scipy/linalg/_decomp.py]: I took the note out of the admonition and made it plain text, and I linked to the appropriate part of the `linalg` tutorial. 

image

I drafted the part of the tutorial I envisioned:

https://github.com/scipy/scipy/blob/1c21193035bc3563cde13a45656375e1c22461c9/doc/source/tutorial/linalg_batch.ipynb

When it looks good, I'll convert it to markdown and link to it from the tutorial and the standard note. With this plan, the documentation identifies the *core shape*s  of the inputs/outputs and is written for a *batch shape* of `()`. This is (much) easier for us, less error-prone, and it is as simple as possible for readers who are interested in the typical use case. For the rest, it is easy to extend to higher-dimensional batches according to the the general rules given in the tutorial.

File: doc/source/tutorial/linalg_batch.ipynb
Comment (mdhaber) [inline doc/source/tutorial/linalg_batch.ipynb:493]:
```suggestion
    "The rules when the shapes are not identical follow logically enough. Each input can have its own batch shape as long as the shapes are broadcastable according to NumPy's [broadcasting rules](https://numpy.org/doc/stable/user/basics.broadcasting.html). The net batch shape is the broadcasted shape of the individual batch shapes, and the shape of each output is the net batch shape plus its core shape."
```

File: doc/source/tutorial/linalg.rst
Comment (mdhaber) [inline doc/source/tutorial/linalg.rst:966]: Probably will just link to the notebook. Ideally we'd convert all of `linalg.rst` to markdown, maybe add a TOC rather than a single masssive page. We could add another section here about the `overwrite_` inputs and include a link in the standard note where applicable.

File: scipy/_lib/_util.py
Comment (mdhaber) [reply scipy/_lib/_util.py:1046]: Ah, the other complication is that the result dtype of `eig` is data-dependent.

```python3
import numpy as np
from scipy import linalg
rng = np.random.default_rng(4932982394)
m = 10

A = rng.random((m, m))
evals, evecs = linalg.eig(A)
print(evals.dtype, evecs.dtype)

A += A.T + np.eye(m) * m
evals, evecs = linalg.eig(A)
print(evals.dtype, evecs.dtype)
```
Perhaps this is someting else to fix, but I think it's a separate issue.  I would prefer not to *also* check each iteration whether we need to reallocate the array with a new dtype. Can we leave it for a follow-up?

**Comment by fancidev** [comment]
Can I check, if user call `eig(a)` where `a` contains more than two dimensions, are the matrices assumed to be in the last two dimensions, or first two dimensions, or otherwise?

**Comment by mdhaber** [comment]
Last two dimensions for consistency with array-API and NumPy. I think one would find it clear enough from the tutorial, but I should probably include that explicitly as in gh-19413.

File: scipy/_lib/_util.py
Comment (lucascolley) [reply scipy/_lib/_util.py:968]: apologies for only a drive-by comment, but is there motivation for using a leading underscore here in the name `_apply_over_batch` (rather than just `apply_over_batch`)?

Comment (mdhaber) [reply scipy/_lib/_util.py:968]: Just to reinforce that it's private. Not as important now that the file names have underscores. but I am not sure if the convention changed. I think the private `xp` utilities in `_array_api` are an exception; it's clear enough that they're private based on the prefix.

Comment (lucascolley) [reply scipy/_lib/_util.py:968]: Right, I think it's okay either way. For `_lib` and other private modules, I think my mental model is roughly 'underscore if only used in this file, no underscore if used by other files', but I'm not sure how accurate that is to the current codebase.

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [inline doc/source/tutorial/linalg_batch.md:14]: Should this have? I saw it in the other notebooks you added.
```suggestion
```{eval-rst}
.. jupyterlite:: ../../_contents/hypothesis_shapiro.ipynb
   :new_tab: True
```
```

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [inline doc/source/tutorial/linalg_batch.md:116]: Can I link to NumPy's broadcasting rules with the normal markdown syntax?

File: scipy/linalg/_decomp.py
Comment (mdhaber) [inline scipy/linalg/_decomp.py:127]:
```suggestion
    The documentation is written assuming array arguments are of specified
```
Otherwise, how does this look to you?

**Review by mdhaber — COMMENTED** [review]
Thanks @melissawm for the video. I think something like that would be very helpful in the documentation!

I had some questions for you here. Also, would you be willing to review this tutorial and the other documentation here?

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [reply doc/source/tutorial/linalg_batch.md:12]: This is particular to this page, because it hasn't been added to any toctree. If it is added to a regular toctree, we don't need this option. But it is informational in that there are a number of options that can be added to the MyST frontmatter - see https://myst-nb.readthedocs.io/en/latest/authoring/text-notebooks.html#notebook-level-metadata

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [reply doc/source/tutorial/linalg_batch.md:15]: Yes, this is equivalent to `.. linalg_batch::` in rst. I did not create it initially because I was able to use the `:doc:` reference when building locally (and doc takes a document file name without extension).

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [reply doc/source/tutorial/linalg_batch.md:14]: This is the way to make the notebook executable and have the "Open in notebook" button appear, yes. If that is desired here, feel free to add - the ipynb file is already created as part of the "convert-notebooks" script in the Makefile.

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [reply doc/source/tutorial/linalg_batch.md:116]: Yes, including intersphinx! So you can do 

```
The rules when the shapes are not identical follow logically. Each input can have its own
batch shape as long as the shapes are broadcastable according to
[NumPy's broadcasting rules](#array-broadcasting-in-numpy).
The net batch shape is the broadcasted shape of the individual batch shapes, and the shape
of each output is the net batch shape plus its core shape.
```

where `array-broadcasting-in-numpy` is the label in the broadcasting guide for NumPy:

https://github.com/numpy/numpy/blob/f3b3dc536e624b20822cd37cd01c15e3767af84a/doc/source/user/basics.broadcasting.rst?plain=1#L2

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:15]: Right - I tried `:doc:` in the previous commit (63ab46f), and it works from the tutorial `.rst`, but not from the docstring of `eig`. Is it supposed to be that way, perhaps because it's easier to resolve the file name from within the `doc` folder?

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:12]: Thanks!

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:116]: Ah, great. Will do.

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [inline doc/source/tutorial/linalg_batch.md:81]: If you want to link to the `linalg.expm` docstring, you can use
```suggestion
Not all linear algebra functions map to scalars. For instance, the {func}`scipy.linalg.expm` function maps from a matrix to a matrix with the same shape.
```
You could also use similar syntax as in rst: 

```
{func}~.linalg.expm`
```
but in this particular case it is ambiguous since we have `scipy.linalg.expm` and also `scipy.sparse.linalg.expm` with the same role (func).

**Review by melissawm — COMMENTED** [review]
Left a couple of minor suggestions. 

One thing to consider is to either have a list (which admittedly may be quickly out-of-date) or have some way of identifying which operations will support the batch mode. 

See https://numpy.org/doc/stable/reference/routines.linalg.html#linear-algebra-on-several-matrices-at-once:

> Several of the linear algebra routines listed above are able to compute results for several matrices at once, if they are stacked into the same array.

> **This is indicated in the documentation via input parameter specifications such as `a : (..., M, M) array_like`.** This means that if for instance given an input array `a.shape == (N, M, M)`, it is interpreted as a “stack” of N matrices, each of size M-by-M.

**Review by mdhaber — COMMENTED** [review]
> One thing to consider is to either have a list (which admittedly may be quickly out-of-date) or have some way of identifying which operations will support the batch mode.

With this decorator, almost all functions can support the behavior beginning in the same release. It might be more practical to create a list of exceptions before the release, but that might not even be needed.

> See https://numpy.org/doc/stable/reference/routines.linalg.html#linear-algebra-on-several-matrices-at-once:

Sure, besides the tutorial and in the functions themselves, we can mention this on the `linalg` API page. Is this part of what you meant?

If you also meant to add the ellipses to the shape documentation: my hesitation is that if this is done, the *whole* docstring should (IMO) be rewritten to reflect that the array can be N-dimensional. Sure, there are some simple changes, like `len(x)` -> `x.shape[-1]`, and `A[i, j]` -> `A[..., i, j]`. But to avoid inconsistencies, it requires very careful consideration of everything the docstring says, so it's error prone, and it would make the essence of the docstring more challenging to understand, as it forces the reader to think about N/N+1/N+2-D even if they are only interested in 0/1/2-D arrays. Since the batch rules are simple and general, they should not affect how one thinks about how a given function works; batching should just be considered a standard feature, and the general documentation is enough to use that feature of any function. The risk with this plan is that people don't know that the feature is available, which we can ameliorate with other forms of advertising it. But in any case, this seems less bad to me than the alternative, which is a lot of work to write less readable and potentially inaccurate documentation. How does that sound?

**Comment by mdhaber** [comment]
Great! This will still be slower than the NumPy versions, so it's OK to mention that in the NumPy tutorial. For one, the SciPy functions sometimes do extra checks (e.g. inverse condition number) which can take some time. Also, this decorator loops in pure Python, so batching is initially just for convenience, not speed. That said, this allows us to speed things up by looping in compiled code one function at a time without making the API inconsistent in the meantime.

Could you suggest the wording you have in mind for linking to the NumPy documentation? Thanks!

**Comment by melissawm** [comment]
Feel free to not add it, but it can be confusing for users to understand that there are two different linalg modules with different features. In the NumPy docs we have this snippet (which will also need to be updated):

> The SciPy library also contains a [linalg](https://docs.scipy.org/doc/scipy/reference/linalg.html#module-scipy.linalg) submodule, and there is overlap in the functionality provided by the SciPy and NumPy submodules. SciPy contains functions not found in [numpy.linalg](https://numpy.org/devdocs/reference/routines.linalg.html#module-numpy.linalg), such as functions related to LU decomposition and the Schur decomposition, multiple ways of calculating the pseudoinverse, and matrix transcendentals such as the matrix logarithm. Some functions that exist in both have augmented functionality in [scipy.linalg](https://docs.scipy.org/doc/scipy/reference/linalg.html#module-scipy.linalg). For example, [scipy.linalg.eig](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eig.html#scipy.linalg.eig) can take a second matrix argument for solving generalized eigenvalue problems. Some functions in NumPy, however, have more flexible broadcasting options. For example, [numpy.linalg.solve](https://numpy.org/devdocs/reference/generated/numpy.linalg.solve.html#numpy.linalg.solve) can handle “stacked” arrays, while [scipy.linalg.solve](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve.html#scipy.linalg.solve) accepts only a single square array as its first argument.

File: doc/source/tutorial/linalg_batch.md
Comment (melissawm) [inline doc/source/tutorial/linalg_batch.md:16]: I think this should be
```suggestion
.. jupyterlite:: ../_contents/linalg_batch.ipynb
```
because we are under `source/tutorials/` and not under `source/tutorials/stats`

**Review by ev-br — COMMENTED** [review]
Looks good in general! Several optional nits inline.

Performance-wise and memory-wise, I second Ilhan's comments about preallocating instead of np.stack-ing and `.append`-ing. I suppose the `argdef` arguments would need to grow some knowledge of what to allocate.

However, whether this is crucial or not depends on what we are using this decorator for. If it's the way to reuse in other linalg functions in production, performance is a concern. If we keep this as a testing implement for lower-level implementations of batching (https://github.com/scipy/scipy/pull/21935#issuecomment-2496094903), then it can just go in as is.

At any rate, I think it's best hold off merging this until after branching of 1.15, and aim to address the whole scipy.linalg in the 1.16 time frame.

File: scipy/_lib/_util.py
Comment (mdhaber) [reply scipy/_lib/_util.py:1044]: @ev-br Can we continue the discussion about pre-allocating memory in this thread?

Again, I considered pre-allocating, but opted against it when implementing this originally because I don't think it is worth the complexity, especially when this is going to be replaced anyway.

Here is a simplified case using `solve` (because there is only one output) to compare the two options.

```python3
import numpy as np
rng = np.random.default_rng(4578345834932546)
n, m = 10000, 100
A = rng.random((n, m, m))
b = rng.random(m)

def preallocate(A, b):
    shape = A.shape[:-1]
    out0 = np.linalg.solve(A[0], b)
    out = np.empty(shape, dtype=out0.dtype)
    out[0] = out0
    for i, Ai in enumerate(A[1:]):
        out[i+1] = np.linalg.solve(Ai, b)
    return out

def stack(A, b):
    return np.stack([np.linalg.solve(Ai, b) for Ai in A])

res = preallocate(A, b)  # 939 ms ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
ref = stack(A, b)  # 955 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Because the time is dominated by the looping and execution of the function, preallocation is not substantially faster. 

The only difference is a momentary factor of 2 on peak memory consumption. (Before the `stack`, the stacking option requires on average half the memory, and after the `stack`, the list of arrays can be garbage collected.)

Note that this `preallocate` version is actually _simpler_ than it would need to be, considering that the output dtype of `eig` is data-dependent: I am only checking the dtype of the output of the first slice, but really, it would need to be checked in each iteration. If the dtype goes from real to complex, we'd have to reallocate, move the existing results over, and pick up where we left off.

Can this PR focus on the interface, documentation, and correctness; and if someone wants to squeeze out that factor of two (before all this code is replaced anyway), that can be done separately?

File: scipy/_lib/_util.py
Comment (mdhaber) [inline scipy/_lib/_util.py:991]: Would this resolve https://github.com/scipy/scipy/pull/21462/files#r1871011248 @ev-br?
```suggestion
    `linalg.eig` accepts two matrices as the first two arguments `a` and `b`, where
    `b` is optional, and returns one array or a tuple of arrays, depending on the
    values of other positional or keyword arguments. To generate a wrapper that applies
    the function over batches of `a` and optionally `b` :
```

File: doc/source/tutorial/linalg_batch.md
Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:61]: I can move the first sentence up to the beginning of the previous section.

For the rest, it sounds like you have an idea for what this admonition would look like. I would be happy to review that if you'd like to submit that in a follow-up PR.

**Review by mdhaber — COMMENTED** [review]
@melissawm Could you take a look at https://github.com/scipy/scipy/pull/21462/files#r1754653523 ~~and the second comment below~~? 

@ev-br Thanks for the review. The rest of the comments are for you.

File: scipy/linalg/tests/test_decomp.py
Comment (ev-br) [reply scipy/linalg/tests/test_decomp.py:409]: I'd guess that working in-place on a batch needs pretty esoteric array strides: the last two dimensions need to be F-ordered:

```
In [12]: from scipy.linalg import inv

In [13]: a = np.arange(9).reshape(3,3) + np.eye(3)

In [14]: np.may_share_memory(a, inv(a, overwrite_a=True))
Out[14]: False

In [15]: np.may_share_memory(a, inv(a.T, overwrite_a=True))
Out[15]: True

```

now for a batch

```
In [26]: b = a.T * np.arange(1, 5)[:, None, None]   # note a.T

In [27]: b.strides
Out[27]: (72, 8, 24)

In [28]: np.may_share_memory(b, inv(b[1, ...], overwrite_a=True))
Out[28]: True

```

File: scipy/_lib/_util.py
Comment (ev-br) [reply scipy/_lib/_util.py:1044]: I think there are two things at play here, First, for large enough core dimensions, the linear algebra big-O complexity dominates; Second, indeed, when looping in python, pre-allocation does not buy much.
To see the perf budget, add a `np.linalg.solve` comparison: for `m=100` your python loop versions are essentially equivalent to numpy, while for `m=10` there's a factor of 5 or so:

```
In [39]: import numpy as np
    ...: rng = np.random.default_rng(4578345834932546)
    ...: n, m = 10000, 10
    ...: A = rng.random((n, m, m))
    ...: b = rng.random(m)
    ...: 
    ...: def preallocate(A, b):
    ...:     shape = A.shape[:-1]
    ...:     out0 = np.linalg.solve(A[0], b)
    ...:     out = np.empty(shape, dtype=out0.dtype)
    ...:     out[0] = out0
    ...:     for i, Ai in enumerate(A[1:]):
    ...:         out[i+1] = np.linalg.solve(Ai, b)
    ...:     return out
    ...: 
    ...: def stack(A, b):
    ...:     return np.stack([np.linalg.solve(Ai, b) for Ai in A])
    ...: 
    ...: %timeit np.linalg.solve(A, b)
    ...: %timeit preallocate(A, b)  # 939 ms ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
    ...: %timeit stack(A, b)  # 955 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
13.9 ms ± 215 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
73.1 ms ± 265 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
73.5 ms ± 303 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Conclusion: as long as looping is in python, extra complexity is not worth it.

File: doc/source/tutorial/linalg_batch.md
Comment (ev-br) [reply doc/source/tutorial/linalg_batch.md:61]: Moving the first sentence would be nice indeed.

For the rest, I think we can roll with what you have and improve iteratively, no need to drag this PR.

File: scipy/_lib/_util.py
Comment (ev-br) [reply scipy/_lib/_util.py:993]: Yup, that's an improvement. Still am not sure how to spell out the difference between one required one optional arg and two required args, but since it's not user-facing and devs are expected to be ready to poke around in code, not just docs, it's fine with https://github.com/scipy/scipy/pull/21462#discussion_r1878399436

File: doc/source/tutorial/linalg_batch.md
Comment (ev-br) [inline doc/source/tutorial/linalg_batch.md:24]: Re-reading this, how about adding a small tl;dr explanation around here + stressing that the following narrative text is a detailed explanation.

Consider adding something similar to

In a nutshell, an array of shape `(4, 3, 2)` can be interpreted as a batch of four 3-by-2 matrices (then we say that the the *core shape* is (3, 2) and the *batch shape* is `(4,)`) or as a `(4,3)`batch of 2D vectors (then the *core shape* is `(2,)` and the *batch shape* is `(4, 3)`). The length of the core shape is also referred to as the *core dimension*.
See below for more details.

Obviously, it's up to you --- if feel it's good as is, fine by me.

Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:24]: Added this to the "Batch Support" section of the main `linalg` tutorial, which is what links to the full tutorial.

Comment (mdhaber) [reply doc/source/tutorial/linalg_batch.md:24]: Did this address your comments @ev-br?

Comment (ev-br) [reply doc/source/tutorial/linalg_batch.md:24]: Yes it did, thank you!

**Review by ev-br — APPROVED** [review]
This now looks good to me. 
The documentation bit is great; the decorator/python looping I think we all agree would be nice to move to compiled code; hopefully this materializes in this release cycle.

So an LGTM from me; won't hit the green button though since it's a linalg enhancement, so I defer to @ilayn

**Comment by ilayn** [comment]
I think the functionality is sound, based on the tests we have here. The remaining details about f2py handling of strides etc. can be tackled later on in my opinion.

**Comment by ilayn** [comment]
Thanks Matt, Melissa and the reviewers.

**Pull request merged**
